routers/web/repo/view.go

```diff
@@ -157,87 +221,12 @@ func localizedExtensions(ext, languageCode string) (localizedExts []string) {
 	return []string{lowerLangCode + ext, ext}
 }
 
-func findReadmeFile(ctx *context.Context, entries git.Entries, treeLink string) (*namedBlob, string) {
-	// Create a list of extensions in priority order
-	// 1. Markdown files - with and without localisation - e.g. README.en-us.md or README.md
-	// 2. Txt files - e.g. README.txt
-	// 3. No extension - e.g. README
-	exts := append(localizedExtensions(".md", ctx.Language()), ".txt", "") // sorted by priority
-	extCount := len(exts)
-	readmeFiles := make([]*namedBlob, extCount+1)
-
-	docsEntries := make([]*git.TreeEntry, 3) // (one of docs/, .gitea/ or .github/)
-	for _, entry := range entries {
-		if entry.IsDir() {
-			lowerName := strings.ToLower(entry.Name())
-			switch lowerName {
-			case "docs":
-				if entry.Name() == "docs" || docsEntries[0] == nil {
-					docsEntries[0] = entry
-				}
-			case ".gitea":
-				if entry.Name() == ".gitea" || docsEntries[1] == nil {
-					docsEntries[1] = entry
-				}
-			case ".github":
-				if entry.Name() == ".github" || docsEntries[2] == nil {
-					docsEntries[2] = entry
-				}
-			}
-			continue
-		}
-
-		if i, ok := markup.IsReadmeFileExtension(entry.Name(), exts...); ok {
-			log.Debug("Potential readme file: %s", entry.Name())
-			name := entry.Name()
-			isSymlink := entry.IsLink()
-			target := entry
-			if isSymlink {
-				var err error
-				target, err = entry.FollowLinks()
-				if err != nil && !git.IsErrBadLink(err) {
-					ctx.ServerError("FollowLinks", err)
-					return nil, ""
-				}
-			}
-			if target != nil && (target.IsExecutable() || target.IsRegular()) {
-				readmeFiles[i] = &namedBlob{
-					name,
-					isSymlink,
-					target.Blob(),
-				}
-			}
-		}
-	}
-
-	var readmeFile *namedBlob
-	readmeTreelink := treeLink
-	for _, f := range readmeFiles {
-		if f != nil {
-			readmeFile = f
-			break
-		}
-	}
-
-	if ctx.Repo.TreePath == "" && readmeFile == nil {
-		for _, entry := range docsEntries {
-			if entry == nil {
-				continue
-			}
-			var err error
-			readmeFile, err = getReadmeFileFromPath(ctx, ctx.Repo.Commit, entry.GetSubJumpablePathName())
-			if err != nil {
-				ctx.ServerError("getReadmeFileFromPath", err)
-				return nil, ""
-			}
-			if readmeFile != nil {
-				readmeFile.name = entry.Name() + "/" + readmeFile.name
-				readmeTreelink = treeLink + "/" + util.PathEscapeSegments(entry.GetSubJumpablePathName())
-				break
-			}
-		}
+func findReadmeFile(ctx *context.Context) (*namedBlob, error) {
```

Comment:
I think the original function wants to avoid read the entries secondly, but the new function will have to read the entries again.

Reply:
Thanks for catching that. I noticed that and thought about it; but in the old version, it was

https://github.com/go-gitea/gitea/blob/6e226057933389710f2451b148faef8ed5ea1ca1/routers/web/repo/view.go#L228

(which is doing e.g. `getReadmeFileFromPath(ctx, ctx.Repo.Commit, "docs")`)

which then recomputed the TreeEntry:

https://github.com/go-gitea/gitea/blob/6e226057933389710f2451b148faef8ed5ea1ca1/routers/web/repo/view.go#L61

so stashing the TreeEntries wasn't optimizing anything, and might even have been slower since `GetSubJumpablePathName()` is itself a recursive function.

Even if it was doing a useful optimization, I was hoping you would accept that both forms are ultimately O(n), and that the benefit of simpler code outweighs the risk of lag. A folder would have to be extremely large for the slowdown to be noticeable.

Reply:
So `getReadmeFileFromPath` or `getReadmeFileInPath` could be refactored to accept a entry list as argument to reduce read the tree again.

Reply:
I took this to suggest passing a `*git.Tree` not `[]*git.TreeEntry`. I think that's probably a more natural signature. I've implemented it in f2acee3d6a5cb1787c45306beda773ebef3faa37.

But it's less elegant this way:

1. I had to go mucking about in the git API because I couldn't find a public method that would let me turn a TreeEntry into a Tree:

    https://github.com/go-gitea/gitea/blob/f2acee3d6a5cb1787c45306beda773ebef3faa37/modules/git/tree_entry.go#L104-L111
2. Because we no longer have an absolute path to inspect, I had to add this `searchSubtrees` flag

    https://github.com/go-gitea/gitea/blob/f2acee3d6a5cb1787c45306beda773ebef3faa37/routers/web/repo/view.go#L98

Is messing with your Git API worth the rare time savings? I'm leaning towards no.

Reply:
@silverwind maybe you could give this a second round of review now that I've changed it to add `TreeEntry.Tree()`.

Reply:
How about rename it as `findReadmeFileInEntries` and replace `*git.Tree` as `[]git.Entry`?

Reply:
I don't understand why that would be better. I've tried it out and it only has accomplished making the code a few lines longer, which you can see in [this diff](https://github.com/neuropoly/gitea/pull/38/files).

Were you hoping that by working in `git.TreeEntry`s it would avoid the need to add `.Tree()`? It doesn't. `modules/web/repo` module doesn't have any way to go `git.TreeEntry` => `.ListEntries()`, because a `git.TreeEntry` doesn't know its contents, it only knows its SHA1, so something needs to call `git.Repository.getTree()`; previously that was being done here

https://github.com/go-gitea/gitea/blob/70d15e77851a437384de018ffebcf4fb3c5af789/modules/git/tree.go#L39

but because `.repo` is package-private, there's no way that the view code can use that line directly.

<details><summary>Tracing</summary>

Optionally, if it would help you follow my reasoning, read this.

The previous code called

https://github.com/go-gitea/gitea/blob/70d15e77851a437384de018ffebcf4fb3c5af789/routers/web/repo/view.go#L228

which calls

https://github.com/go-gitea/gitea/blob/70d15e77851a437384de018ffebcf4fb3c5af789/modules/git/tree_entry.go#L109

which calls

https://github.com/go-gitea/gitea/blob/70d15e77851a437384de018ffebcf4fb3c5af789/modules/git/tree.go#L39

So now I'm calling

https://github.com/neuropoly/gitea/blob/b02f0611e38e09fd93ebbb6d8d53cde014b14ece/routers/web/repo/view.go#L170

directly which does the same thing in the end

https://github.com/go-gitea/gitea/blob/b02f0611e38e09fd93ebbb6d8d53cde014b14ece/modules/git/tree_entry.go#L106

but with less steps in between.

</details>

Reply:
`findReadmeFile` could also accept entries which has been loaded in `renderDirectory`.

Reply:
I see. I thought when you said "avoid read entries secondly" you were talking about the `.ListEntries` call in `FindReadmeFile`, but you were actually talking about

https://github.com/go-gitea/gitea/blob/47efba78ec966631ccdea3e05a50f4cf59ca9fd3/routers/web/repo/view.go#L783-L804

(I didn't see what the problem was at first because in the old code, it _was_ reading entries a second time:

https://github.com/go-gitea/gitea/blob/47efba78ec966631ccdea3e05a50f4cf59ca9fd3/routers/web/repo/view.go#L60-L66

but that was the cold path that only happened if someone was keeping their README in e.g. `.gitea/README.md` folder, so it wasn't a big deal)

I've put back the parts from `renderDirectoryFile` that I took out and now I think it makes sense. Thank you for your guidance.

Reply:
Okay I think this subthread is done now? Unresolve it if you disagree.